### PR TITLE
[bug] fix shared library AddressSanitizer integration for the gnu compiler

### PR DESCRIPTION
**Pull request by icfaust**
## Description

The logic was getting ugly for integrating ASan, and as compiler/linker integrations of AddressSanitizer are under development (see https://devblogs.microsoft.com/cppblog/addresssanitizer-asan-for-windows-with-msvc/).  Related linkers have different default settings and different available options, this moves them to the compiler makefile definitions in order to isolate changes and maintain the central logic.

GNU compilers don't have the -shared-libasan linker flag, while clang, icx and dpcpp do. Windows doesn't maintain any of these flags (this is still very untested, but is just trying to simplify the work when verifying its operation at a later date).

Note this also doesn't fix some issues with other sanitizers, which for Clang will default to static linking (making introspection between the oneDAL shared librarys difficul/impossible) or for GNU which will default to shared linking.  This does not touch those issues, and is necessary for fixing ASan specifically. Some compilers/linkers don't allow for these other sanitizers at all.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

**Comment by david-cortes-intel** [comment]
@icfaust This results in a compilation error for the .a files:
```
ar: creating __work_gnu/md/lnx32e/daal/lib/libonedal.a
collect2: error: ld returned 1 exit status
make: *** [makefile:520: __work_gnu/md/lnx32e/daal/lib/libonedal_core.so] Error 1
```

**Comment by david-cortes-intel** [comment]
The error appears to be something along these lines:
```
/usr/bin/ld: cannot find /usr/lib64/libasan.so.6.0.0
```

The available file in this system would be as follows:
```
/usr/lib/gcc/x86_64-redhat-linux/11/libasan.so
```

Not sure what exactly is making this PR look for this different filename.

File: dev/make/compiler_definitions/gnu.mk [hunk 1]
```diff
@@ -28,6 +28,9 @@ CORE.SERV.COMPILER.gnu = generic
 -Zl.gnu =
 -DEBC.gnu = -g
 
+-asanstatic.gnu = -static-libasan
```
Comment (david-cortes-intel) [inline dev/make/compiler_definitions/gnu.mk:31]:
```suggestion
-asanstatic.gnu =
```

It at least compiles successfully if this argument is removed. Not sure about the impact though.

Comment (icfaust) [reply dev/make/compiler_definitions/gnu.mk:31]: Thanks for checking, it looks like by not including it, it will default to dynamic linking (https://gcc.gnu.org/onlinedocs/gcc/Link-Options.html#index-static-libasan).  Seems like an issue with GCC, should we leave a note and remove it, saying that static linking with GNU compilers is temporarily unavailable?

Comment (icfaust) [reply dev/make/compiler_definitions/gnu.mk:31]: By remove I mean take your suggestion.

Comment (david-cortes-intel) [reply dev/make/compiler_definitions/gnu.mk:31]: > Thanks for checking, it looks like by not including it, it will default to dynamic linking (https://gcc.gnu.org/onlinedocs/gcc/Link-Options.html#index-static-libasan). Seems like an issue with GCC, should we leave a note and remove it, saying that static linking with GNU compilers is temporarily unavailable?

I cannot think of any usage for a .a version of oneDAL with ASAN, so yes maybe it could added as a note.

Although on a second look, this appears to be a redhat-specific issue with misconfigured software, and I have a feeling it might be something specific to the way our servers are set up.

Comment (david-cortes-intel) [reply dev/make/compiler_definitions/gnu.mk:31]: @icfaust After some experiments with my setup, turns out it's not strictly necessary to remove this line.

But I still don't think the resulting file would be of much use - for example, in the case of sklearnex, if one tries to use that .a file, one might also want to build sklearnex itself with asan, but it cannot link statically to it (I think it's a limitation of cython but not 100% sure).

**Comment by icfaust** [comment]
> @icfaust PR has conflicts with the main branch in the makefile now.

@david-cortes-intel Could you give some more info on this? (i.e. a reproducer) As of now, it merged main without issue, so some guidance will save some time in fixing.
